pkg/apis/core/validation/validation.go

```diff
@@ -3252,10 +3288,11 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
 // This includes checking formatting and uniqueness.  It also canonicalizes the
 // structure by setting default values and implementing any backwards-compatibility
 // tricks.
-func ValidatePodSpec(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+// The pod metadata is needed to validate generic ephemeral volumes. It is optional.
+func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	vols, vErrs := ValidateVolumes(spec.Volumes, fldPath.Child("volumes"))
+	vols, vErrs := ValidateVolumes(spec.Volumes, podMeta, fldPath.Child("volumes"))
```

Comment:
Hrm.... I don't know that we should be validating volume names in PodSpec in StatefulSets, Daemonsets, or ReplicaSets (or in other workload sets) because we have no idea what pod name a statefulset is going to choose based on the name.

Instead, I think this validation should only be for Pods created directly for now.  Jobs, for instance, have their own name rules.

Reply:
`podMeta` will be empty when validating PodSpec in those other scenarios, in which case the name validation is disabled. So that part should be fine.

However, when adding the nil parameter I unintentionally also added it in `validatePodMetadataAndSpec` instead of passing the real podMeta of the pod. Fixed.

That this slipped through testing showed that the new tests were too low-level. Therefore I moved the test cases into the more complete `TestValidatePod`. While doing so, I learned that the generated name is set already as `podMeta.Name`. That simplifies the checking because it can rely on that value.

I couldn't resist modernizing those tests by using `t.Run` and named test cases.